### PR TITLE
refactor: use v8 serialization for ipc

### PR DESCRIPTION
**Pull request by nornagon**
#### Description of Change

BREAKING CHANGE

This changes IPC communication to use v8's [Structured Clone](https://html.spec.whatwg.org/multipage/structured-data.html#safe-passing-of-structured-data) algorithm instead of using the `base::Value` serialization defined in `native_mate_converters/v8_value_converter.cc`. This is be faster, more featureful, and less surprising than the existing logic, since it more-or-less matches the logic that backs `postMessage`.

This brings about a 2x performance boost for large buffers and complex objects. Latency for small messages is not significantly affected.

User-observable differences from the existing IPC API:
- It's about 2x faster.
- `NaN`, `Infinity`, and `undefined` are transferred as such, rather than being converted to `null`.
- Cyclic objects can be transmitted.
- `Set` and `Map`, `Error`, `RegExp`, `Date` and `BigInt` can be transmitted.
- `Buffer` will be converted to `Uint8Array`.
- Typed arrays (`Float32Array` and friends) will be transmitted as they are, instead of being converted to `Buffer`.
- Sparse arrays will be transferred as sparse arrays, instead of being converted to dense arrays with `null`s.

> **NOTE**: Objects that aren't serializable with V8's Structured Clone algorithm, such as functions, DOM objects, special Node/Electron objects like `process.env` or `WebContents`, or any objects containing such items will be serialized with the old `base::Value`-based algorithm. However, this behavior is deprecated and will throw an exception beginning with Electron 9.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: IPC between main and renderer processes now uses the Structured Clone Algorithm.

### CODE REVIEW

File: lib/browser/chrome-extension.js [hunk 1]
```diff
@@ -240,7 +240,7 @@ const getMessagesPath = (extensionId) => {
 
 ipcMainUtils.handleSync('CHROME_GET_MESSAGES', async function (event, extensionId) {
   const messagesPath = getMessagesPath(extensionId)
-  return fs.promises.readFile(messagesPath)
+  return fs.promises.readFile(messagesPath).then(b => b.toString('utf8'))
```
Comment (MarshallOfSound) [inline lib/browser/chrome-extension.js]: Can we not send buffers over IPC anymore??

Comment (nornagon) [reply lib/browser/chrome-extension.js]: you totally can, but there's no particular reason to here, and the remote end will end up as a Uint8Array, which doesn't have this form of `toString()`.

**Comment by nornagon** [comment]
Remaining TODOs as of latest commit:

- [ ] Handle errors gracefully & correctly.
- [ ] Investigate whether we can discriminate between `Buffer` and `Uint8Array` wisely, and deserialize `Buffer`s as `Buffer`s where we can, and `Uint8Array` where we can't (e.g. in sandboxed renderers).
- [ ] Convert the remaining methods to the new serialization algorithm (`webContents.send`, `messageTo`, and `messageHost`).

**Comment by nornagon** [comment]
OK, there's a choice to make about how we handle when non-POJO objects are passed to IPC functions (things like DOM objects, `WebContents`, `process.env`, etc.).

1. Throw an exception.
    - This is the safest option, but also the most disruptive. (e.g. many of Electron's internal APIs pass such objects to IPC methods, and will need to be refactored.)
2. JSON.stringify()/JSON.parse() them.
    - This is roughly similar to current behavior of the IPC methods, in that it can't handle NaN or Infinity, or cyclic objects. However, it might be somewhat confusing to have some objects work differently based on whether they're referenced from a non-POJO object or not. It's a fairly obscure corner case though, so maybe nobody would notice?
3. Iterate the keys/values of such objects and recurse in the serialization algorithm.
    - This gives us cyclic object support and correct handling of Infinity/NaN/etc.
    - But most non-trivial non-POJO objects contain unserializable things like functions.
    - We can attempt to replace those unserializable things with `undefined`, but it only works one layer deep. i.e. if there's a non-POJO object like `{ fn: () => {} }`, it'll be turned into `{ fn: undefined }`, but if it's like `{ events: { fn: () => {} } }`, then the entire serialization operation must be aborted and an error thrown.

I'm leaning towards (1), since I don't think it really makes sense to be trying to send `window.location` or a `WebContents` value over IPC, and places where we're doing that are likely error-prone. Thoughts?

**Comment by MarshallOfSound** [comment]
I don't like the idea of existing code throwing errors in an obscure and un-lintable way.  i.e. we can't provide a lint rule or other tool to detect invalid usages of `ipc.send` so folks will only find out at runtime when it's too late and hard to spot.

I'm leaning towards something equivalent to (2) where if the v8 serializer fails we fall back to the old one.  Maybe log a console warning that you tried to send an unserializable object and we fell back to a less accurate converter?

**Comment by nornagon** [comment]
Does that leave any path towards eventually eliminating the old converter? I'm disinclined to leave it there forever... but maybe it would make sense as a less-dramatic deprecation strategy? e.g. "Looks like you sent a non-standard object over IPC. This is probably not what you intended to do, but we obliged and serialized it anyway. Look out, this will throw an exception in Electron 9."

**Comment by MarshallOfSound** [comment]
I'd be down for a **very** long term deprecation strategy here.  Like we should support serialization for I'd say 4 versions beyond when this lands (number picked to match the number of versions we used for the native module deprecation).

The deprecation log should include the object itself as well if possible 🤔

**Comment by nornagon** [comment]
@MarshallOfSound I'm not sure that's warranted. 99% of cases where IPC is used, the new version should still work fine. Only in weird cases where it only kind of semi-works currently are problematic.

For example. The `new-window` event has as one of its arguments the `options` parameter, which is the options passed to `new BrowserWindow()`. One of these options is `webContents`, which specifies which WebContents will go in the new window. For WebViews, the `new-window` event is proxied over to the host frame. However, during that transmission, the `webContents` key stops being an actual `WebContents`. Is anything depending on that key being there in the webview host? I don't think so. Can anything actually useful be done with it there? Again, I don't think so. The fact that it's there is an accident, and it never really performed any kind of useful function.

Also, it's trivial to get something close to the old behavior with the transformation `x => JSON.parse(JSON.stringify(x))`, as long as `x` isn't cyclic.

tldr most people wouldn't need to fix anything if this usage started throwing an error, and for those that do, the fix is easy.

**Comment by MarshallOfSound** [comment]
> if this sort of usage started throwing an error

My only concern here was that you can't glance at a bit of code and say "that's gonna cause a problem".  In fact I don't think you could describe the kinds of things that would cause this problem very easily at all.  That's the only reason I'm advocating for a deprecation cycle, it will take time for people to find all the edge cases in their app that might do this.  Maybe 4 versions was a bit over-eager 🤷‍♂But it should definitely go through a cycle where the old code still works but tells you you're doing something bad.

**Comment by nornagon** [comment]
The things which will throw exceptions with structured clone that don't with base::Value serialization:
- DOM objects (`Element`, `Location`, etc.)
- special Node objects like `process.env` or http internals
- special Electron API objects like WebContents
- functions

I've updated the PR to fall back to the base::Value-based serialization if structured clone fails, and emit a deprecation warning.

**Comment by nornagon** [comment]
The last remaining tricky thing is that sending a `Buffer` will result in the other side receiving a `Uint8Array`. Without patching V8, I don't see a good way to get around this. Fortunately, you can wrap a `Buffer` around a `Uint8Array` quite easily:

```
const uint8Array = await ipcRenderer.invoke('return-a-buffer')
const buf = Buffer.from(uint8Array.buffer)
```

**Comment by nornagon** [comment]
Taking off the [wip] as I believe this is ready for full review.

File: shell/renderer/api/atom_api_renderer_ipc.cc [hunk 2]
```diff
@@ -69,45 +70,71 @@ class IPCRenderer : public gin::Wrappable<IPCRenderer> {
...
 
   v8::Local<v8::Promise> Invoke(v8::Isolate* isolate,
                                 bool internal,
                                 const std::string& channel,
-                                const base::Value& arguments) {
-    electron::util::Promise<base::Value> p(isolate);
+                                v8::Local<v8::Value> arguments) {
+    blink::CloneableMessage message;
+    if (!mate::ConvertFromV8(isolate, arguments, &message)) {
+      return v8::Local<v8::Promise>();
```
Comment (MarshallOfSound) [inline shell/renderer/api/atom_api_renderer_ipc.cc]: Is this promise rejected or does the method throw an error and return an empty promise?

Comment (nornagon) [reply shell/renderer/api/atom_api_renderer_ipc.cc]: When the serialization fails, the method throws an error. Also, this isn't a `Promise` value, it's an empty `Local`.

Comment (MarshallOfSound) [reply shell/renderer/api/atom_api_renderer_ipc.cc]: My concern here was is that error still catchable with .catch?

If it's an empty promise and the method throws then .catch won't work which is incompatible with a promise based API

Comment (nornagon) [reply shell/renderer/api/atom_api_renderer_ipc.cc]: yes, it's still catchable.

```js
ipcRenderer.invoke('foo', ()=>{}).catch(e => console.log('bloop'))
// => bloop
```

File: patches/chromium/export_fetchapi_mojo_traits_to_fix_component_build.patch [hunk 3]
```diff
@@ -0,0 +1,36 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jeremy Apthorp <jeremya@chromium.org>
+Date: Fri, 20 Sep 2019 16:44:18 -0400
+Subject: export FetchAPI mojo traits to fix component build
+
+Without these, we get link errors in the component build when using the
+blink::CloneableMessage mojo traits.
```
Comment (deepak1556) [inline patches/chromium/export_fetchapi_mojo_traits_to_fix_component_build.patch]: Is there a CL for this ?

File: docs/api/breaking-changes.md [hunk 4]
```diff
@@ -6,6 +6,59 @@ Breaking changes will be documented here, and deprecation warnings added to JS c
 
 The `FIXME` string is used in code comments to denote things that should be fixed for future releases. See https://github.com/electron/electron/search?q=fixme
 
+## Planned Breaking API Changes (8.0)
```
Comment (jkleinsc) [inline docs/api/breaking-changes.md]: Are we targeting this for 8 or 9?  The PR says 9.

Comment (nornagon) [reply docs/api/breaking-changes.md]: This PR is targeted to 8. When 9 rolls around, the behavior will change again (the deprecated things will start throwing an exception instead of being serialized with the old algorithm).

Comment (jkleinsc) [reply docs/api/breaking-changes.md]: @nornagon doesn't that mean it will be `Planned Breaking API Changes (9.0)` in the doc?

Comment (jkleinsc) [reply docs/api/breaking-changes.md]: Ehh... nevermind I misread this.

File: patches/chromium/export_fetchapi_mojo_traits_to_fix_component_build.patch (hunk 3 above)
Comment (nornagon) [reply patches/chromium/export_fetchapi_mojo_traits_to_fix_component_build.patch]: Not yet, but I talked to pfeldman@ and he said he'd approve it if i uploaded it :P will do soon.

File: lib/browser/guest-view-manager.js [hunk 5]
```diff
@@ -48,6 +47,13 @@ const supportedWebViewEvents = [
 const guestInstances = {}
 const embedderElementsMap = {}
 
+function sanitizeOptionsForGuest (options) {
+  const ret = { ...options }
```
Comment (miniak) [inline lib/browser/guest-view-manager.js]: what about?
```js
const { webContents, ...ret } = options
return ret
```

File: lib/browser/chrome-extension.js (hunk 1 above)
Comment (miniak) [inline lib/browser/chrome-extension.js]: what about?
```js
const buffer = await fs.promises.readFile(messagesPath)
return toString('utf8')
```

Comment (miniak) [inline lib/browser/chrome-extension.js]: you can do just
```js
return fs.promises.readFile(messagesPath, 'utf8')
```

File: lib/browser/guest-view-manager.js (hunk 5 above)
Comment (miniak) [inline lib/browser/guest-view-manager.js]: what about?
```js
// WebContents values can't be sent over IPC.
const { webContents, ...ret } = options
return ret
```

File: shell/common/api/api.mojom [hunk 6]
```diff
@@ -1,19 +1,25 @@
...
       bool internal,
       bool send_to_all,
       string channel,
-      mojo_base.mojom.ListValue arguments,
+      blink.mojom.CloneableMessage arguments,
       int32 sender_id);
 
   UpdateCrashpadPipeName(string pipe_name);
 
+  // This is an API specific to the "remote" module, and will ultimately be
```
Comment (miniak) [inline shell/common/api/api.mojom]: `#if BUILDFLAG(ENABLE_REMOTE_MODULE)`

File: shell/browser/api/atom_api_web_contents.cc [hunk 7]
```diff
@@ -1047,24 +1047,35 @@ void WebContents::MessageTo(bool internal,
...
 
 void WebContents::MessageHost(const std::string& channel,
-                              base::Value arguments) {
+                              blink::CloneableMessage arguments) {
   // webContents.emit('ipc-message-host', new Event(), channel, args);
   EmitWithSender("ipc-message-host", bindings_.dispatch_context(),
                  base::nullopt, channel, std::move(arguments));
 }
 
+void WebContents::DereferenceRemoteJSObject(const std::string& context_id,
```
Comment (miniak) [inline shell/browser/api/atom_api_web_contents.cc]: #if BUILDFLAG(ENABLE_REMOTE_MODULE)

File: shell/common/api/api.mojom [hunk 8]
```diff
@@ -64,11 +70,18 @@ interface ElectronBrowser {
...
     string channel,
-    mojo_base.mojom.ListValue arguments);
+    blink.mojom.CloneableMessage arguments);
 
   MessageHost(
     string channel,
-    mojo_base.mojom.ListValue arguments);
+    blink.mojom.CloneableMessage arguments);
+
+  // This is an API specific to the "remote" module, and will ultimately be
```
Comment (miniak) [inline shell/common/api/api.mojom]: #if BUILDFLAG(ENABLE_REMOTE_MODULE)

File: shell/renderer/electron_api_service_impl.cc [hunk 9]
```diff
@@ -168,12 +172,36 @@ void ElectronApiServiceImpl::Message(bool internal,
...
         v8::Local<v8::Context> child_context =
             renderer_client_->GetContext(child->ToWebLocalFrame(), isolate);
-        EmitIPCEvent(child_context, internal, channel, arguments.GetList(),
-                     sender_id);
+        EmitIPCEvent(child_context, internal, channel, args, sender_id);
       }
   }
 }
 
+void ElectronApiServiceImpl::DereferenceRemoteJSCallback(
```
Comment (miniak) [inline shell/renderer/electron_api_service_impl.cc]: #if BUILDFLAG(ENABLE_REMOTE_MODULE)

File: shell/renderer/electron_api_service_impl.h [hunk 10]
```diff
@@ -28,8 +28,10 @@ class ElectronApiServiceImpl : public mojom::ElectronRenderer,
   void Message(bool internal,
                bool send_to_all,
                const std::string& channel,
-               base::Value arguments,
+               blink::CloneableMessage arguments,
                int32_t sender_id) override;
+  void DereferenceRemoteJSCallback(const std::string& context_id,
```
Comment (miniak) [inline shell/renderer/electron_api_service_impl.h]: #if BUILDFLAG(ENABLE_REMOTE_MODULE)

File: docs/api/breaking-changes.md [hunk 11]
```diff
@@ -6,6 +6,59 @@ Breaking changes will be documented here, and deprecation warnings added to JS c
...
+  arrays with `null`s.
+- `Date` objects will be transferred as `Date` objects, instead of being
+  converted to their ISO string representation.
+- Typed Arrays (such as `Uint8Array`, `Uint16Array`, `Uint32Array` and so on)
+  will be transferred as such, instead of being converted to Node.js `Buffer`.
+- Node.js `Buffer` objects will be transferred as `Uint8Array`s. You can
+  convert a `Uint8Array` back to a Node.js `Buffer` by wrapping the underlying
+  `ArrayBuffer`:
+```js
+Buffer.from(someUint8Array.buffer)
```
Comment (miniak) [inline docs/api/breaking-changes.md]: it needs to be `Buffer.from(value.buffer, value.byteOffset, value.byteLength)`

File: docs/api/breaking-changes.md [hunk 12]
```diff
@@ -6,6 +6,59 @@ Breaking changes will be documented here, and deprecation warnings added to JS c
...
+
+// From Electron 8:
+ipcRenderer.send('channel', { value: 3, someFunction: () => {} })
+// => throws Error("() => {} could not be cloned.")
+```
+- `NaN`, `Infinity` and `-Infinity` will now be correctly serialized, instead
+  of being converted to `null`.
+- Objects containing cyclic references will now be correctly serialized,
+  instead of being converted to `null`.
+- `Set`, `Map`, `Error` and `RegExp` values will be correctly serialized,
```
Comment (miniak) [inline docs/api/breaking-changes.md]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
> Error and Function objects cannot be duplicated by the structured clone algorithm; attempting to do so will throw a DATA_CLONE_ERR exception.

File: docs/api/breaking-changes.md (hunk 11 above)
Comment (nornagon) [reply docs/api/breaking-changes.md]: Doesn't seem like it:

```
> u8b = Uint8Array.from([1,2,3])
Uint8Array [ 1, 2, 3 ]
> Buffer.from(u8b.buffer)
<Buffer 01 02 03>
```

Comment (nornagon) [reply docs/api/breaking-changes.md]: oh i see, you mean for when it's a view on a larger buffer. sure.

File: docs/api/breaking-changes.md (hunk 12 above)
Comment (nornagon) [reply docs/api/breaking-changes.md]: DevTools_-_github_com_electron_electron_pull_20214

File: spec/webview-spec.js [hunk 13]
```diff
@@ -1068,7 +1068,7 @@ describe('<webview> tag', function () {
       await loadWebView(webview, { src })
 
       const data = await webview.printToPDF({})
-      expect(data).to.be.an.instanceof(Buffer).that.is.not.empty()
+      expect(data).to.be.an.instanceof(Uint8Array).that.is.not.empty()
```
Comment (miniak) [inline spec/webview-spec.js]: the documentation for `<webview>.printToPDF(options)` should specify the return value as `Promise<Uint8Array >` instead of `Promise<Buffer>`

File: docs/api/breaking-changes.md (hunk 12 above)
Comment (miniak) [inline docs/api/breaking-changes.md]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
> Error and Function objects cannot be duplicated by the structured clone algorithm; attempting to do so will throw a DATA_CLONE_ERR exception.

File: docs/api/breaking-changes.md (hunk 11 above)
Comment (miniak) [inline docs/api/breaking-changes.md]: it needs to be `Buffer.from(value.buffer, value.byteOffset, value.byteLength)` as the underlying `ArrayBuffer` can be larger with some invalid data beyond the current view

**Comment by miniak** [comment]
this will allow us to also simplify the serialization in the `remote` module. #20427

File: shell/common/api/api.mojom (hunk 8 above)
Comment (nornagon) [reply shell/common/api/api.mojom]: Mojom files can't have `#if` in them unfortunately. I'm not sure it's worth splitting this out into an optionally-offered interface, but I can make the implementations `NOTREACHED()`.

File: lib/browser/guest-view-manager.js (hunk 5 above)
Comment (nornagon) [reply lib/browser/guest-view-manager.js]: I think that'd cause an 'unused variable' warning...

File: docs/api/breaking-changes.md (hunk 12 above)
Comment (nornagon) [reply docs/api/breaking-changes.md]: I'm not sure why you deleted my previous comment displaying a screenshot of this working, but here it is again:

DevTools_-_chrome-search___local-ntp_local-ntp_html

File: shell/common/api/api.mojom (hunk 8 above)
Comment (deepak1556) [reply shell/common/api/api.mojom]: https://cs.chromium.org/chromium/src/mojo/public/tools/bindings/mojom.gni?l=274 can help build optional features into mojom files.

File: spec/webview-spec.js (hunk 13 above)
Comment (nornagon) [reply spec/webview-spec.js]: Unfortunately this currently produces an error in our typings generator:

```
../electron.d.ts(10212,62): error TS2694: Namespace 'Electron' has no exported member 'Uint8Array'.
```

Comment (nornagon) [reply spec/webview-spec.js]: https://github.com/electron/typescript-definitions/pull/156 fixes

File: docs/api/breaking-changes.md (hunk 12 above)
Comment (miniak) [reply docs/api/breaking-changes.md]: @nornagon that must have been a mistake, sorry for that. I was deleting my own comments as I've added them as single comments first, but then switched them to review comments.

File: spec-main/api-ipc-renderer-spec.ts [hunk 14]
```diff
@@ -104,7 +105,7 @@ describe('ipcRenderer module', () => {
       expect(childValue).to.deep.equal(child)
     })
 
-    it('inserts null for cyclic references', async () => {
+    it('can handle cyclic references', async () => {
```
Comment (ckerr) [inline spec-main/api-ipc-renderer-spec.ts]: :tada:

File: lib/browser/rpc-server.js [hunk 15]
```diff
@@ -114,7 +114,7 @@ ipcMainUtils.handleSync('ELECTRON_BROWSER_SANDBOX_LOAD', async function (event)
     process: {
       arch: process.arch,
       platform: process.platform,
-      env: process.env,
+      env: { ...process.env },
```
[the remark this answers is not in the document]
Comment (nornagon) [reply lib/browser/rpc-server.js]: `process.env` is a C++-backed "special" JS object, so it can't be serialized by Structured Clone. This code would work unmodified currently, but it would fall back on the old `base::Value` serialization and emit a deprecation warning, and begin throwing an exception in Electron 9.

File: shell/browser/api/atom_api_web_contents.cc [hunk 16]
```diff
@@ -1017,7 +1017,7 @@ void WebContents::OnElectronBrowserConnectionError() {
 
 void WebContents::Message(bool internal,
                           const std::string& channel,
-                          base::Value arguments) {
+                          blink::CloneableMessage arguments) {
```
[the remark this answers is not in the document]
Comment (nornagon) [reply shell/browser/api/atom_api_web_contents.cc]: `blink::CloneableMessage`'s copy constructor is deleted, which ensures that this is called with `std::move` or `CloneableMessage::ShallowClone`.

File: shell/common/native_mate_converters/blink_converter.cc [hunk 17]
```diff
@@ -527,4 +532,182 @@ bool Converter<network::mojom::ReferrerPolicy>::FromV8(
...
+  explicit V8Serializer(v8::Isolate* isolate,
+                        bool use_old_serialization = false)
+      : isolate_(isolate),
+        serializer_(isolate, this),
+        use_old_serialization_(use_old_serialization) {}
+  ~V8Serializer() override = default;
+
+  bool Serialize(v8::Local<v8::Value> value, blink::CloneableMessage* out) {
+    serializer_.WriteHeader();
+    if (use_old_serialization_) {
```
[the remark this answers is not in the document]
Comment (nornagon) [reply shell/common/native_mate_converters/blink_converter.cc]: It is used, see L564 in this file. It will be deleted, along with the serialization tags, in Electron 9.

File: shell/common/native_mate_converters/blink_converter.cc [hunk 18]
```diff
@@ -527,4 +532,182 @@ bool Converter<network::mojom::ReferrerPolicy>::FromV8(
...
+
+  void WriteTag(uint8_t tag) { serializer_.WriteRawBytes(&tag, 1); }
+
+  // v8::ValueSerializer::Delegate
+  void* ReallocateBufferMemory(void* old_buffer,
+                               size_t size,
+                               size_t* actual_size) override {
+    data_.resize(size);
+    *actual_size = data_.capacity();
+    return data_.data();
```
[the remark this answers is not in the document]
Comment (nornagon) [reply shell/common/native_mate_converters/blink_converter.cc]: Chromium (and thus Electron) is built with `-fno-exceptions`, so unfortunately I don't think we can catch the error here. I'm not totally sure what the stdlib's behavior will be when there's an exception thrown during vector resize...

File: shell/renderer/api/atom_api_renderer_ipc.cc [hunk 19]
```diff
@@ -69,45 +70,71 @@ class IPCRenderer : public gin::Wrappable<IPCRenderer> {
...
-  void Send(bool internal,
+  void Send(v8::Isolate* isolate,
+            bool internal,
             const std::string& channel,
-            const base::ListValue& arguments) {
-    electron_browser_ptr_->get()->Message(internal, channel, arguments.Clone());
+            v8::Local<v8::Value> arguments) {
+    blink::CloneableMessage message;
+    if (!mate::ConvertFromV8(isolate, arguments, &message)) {
+      return;
```
[the remark this answers is not in the document]
Comment (nornagon) [reply shell/renderer/api/atom_api_renderer_ipc.cc]: In Electron 7 this will only be hit if both the V8 serializer and the `base::Value` serializer fail. I'm not exactly sure what circumstances the `base::Value` serializer can fail in, but those ones. In Electron 8 this will throw.

The `EmitDeprecationWarning` is taken care of by the `base::Value` path in the serializer.

File: docs/api/ipc-renderer.md [hunk 20]
```diff
@@ -69,9 +75,15 @@ The main process handles it by listening for `channel` with the
 
 Returns `Promise<any>` - Resolves with the response from the main process.
 
-Send a message to the main process asynchronously via `channel` and expect an
-asynchronous result. Arguments will be serialized as JSON internally and
-hence no functions or prototype chain will be included.
+Send a message to the main process via `channel` and expect a result
```
[the remark this answers is not in the document]
Comment (nornagon) [reply docs/api/ipc-renderer.md]: I don't think the distinction is worth the grammatical awkwardness, plus there are certainly elements of the sending that happen synchronously (i.e. the serialization).

File: spec/webview-spec.js (hunk 13 above)
Comment (nornagon) [reply spec/webview-spec.js]: electron/typescript-definitions#157 _actually_ fixes.

File: shell/common/native_mate_converters/blink_converter.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (nornagon) [reply shell/common/native_mate_converters/blink_converter.cc]: I asked about this in #cxx on the Chromium workspace and it looks like the the exception, if thrown, will crash the process. That seems like reasonable behavior to me, so I'm going to resolve this.

File: lib/browser/rpc-server.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (ckerr) [reply lib/browser/rpc-server.js]: Sounds good, thanks for the explanation. I didn't realize `process.env` was a special case compared to other objects.

**Comment by Xylobol** [comment]
For anyone else interested, these changes have landed in nightly - https://github.com/electron/nightlies/releases/tag/v8.0.0-nightly.20191009

**Comment by gdavidkov** [comment]
Will this be backported to 6.x?

**Comment by nornagon** [comment]
No, since it is a breaking change.

**Comment by Slapbox** [comment]
This PR is noted as a **_Breaking Changes_** for Electron 8.0.0 & 9.0.0 betas.  Is there any difference in the IPC between versions 8 and 9?

**Comment by pushkin-** [comment]
@Slapbox  I believe the plan in 8 is to use the new algorithm, but if it fails, emit a warning and fallback to the old algorithm.

In Electron 9, they're going remove the fallback and warning and serializing bad data will start throwing exceptions.

**Comment by Qrysto** [comment]
Is there any way to check if a variable is safe to be transferred through RPC calls? Could `JSON.stringify` be used to test it?

**Comment by nornagon** [comment]
@Qrysto The most reliable way to check for serializability is to try sending the message, and catching the error. If for some odd reason you need to check in a way that's guaranteed not to send a message, you could try sending to a dummy channel that has no listeners, or if you're in the main process or an unsandboxed renderer, you can try the [v8.serialize Node.JS API](https://nodejs.org/api/v8.html#v8_v8_serialize_value). `JSON.stringify` isn't sufficient to determine serializability.
